### ISSUE TITLE
Locking dependencies for production releases

### ISSUE THREAD

**Issue by JoelLarson** [bundler/bundler#5869]
It would be really nice to have a feature that will allow a Gemfile.lock to represent the exact versions to install when being built on a build server.

Bundler currently seems to lock dependencies based on what is defined in your Gemfile and the dependency Gemfiles (and their children): https://github.com/rails/rails/blob/master/Gemfile.lock

While developing, it isn't a problem when the lock contains locks like this:
```
    rails (5.2.0.alpha)
      actioncable (= 5.2.0.alpha)
      actionmailer (= 5.2.0.alpha)
      actionpack (= 5.2.0.alpha)
      actionview (= 5.2.0.alpha)
      activejob (= 5.2.0.alpha)
      activemodel (= 5.2.0.alpha)
      activerecord (= 5.2.0.alpha)
      activesupport (= 5.2.0.alpha)
      bundler (>= 1.3.0)
      railties (= 5.2.0.alpha)
      sprockets-rails (>= 2.0.0)
```

However, this can be problematic when a dependency from a 3rd party decides to update and has a breaking change in a minor or patch version. These types of changes are hard to debug on production servers in specific scenarios.

I would like to propose a way to define some measures to improve on the package security of bundler and install only a single version of code.

`bundle install` would read the `Gemfile.lock` and only look at the exact versions that are defined in it, it would retrieve the version from the repository and asset the checksum matches what would be in the `Gemfile.lock` to prevent mysterious problems caused by people republishing versions, and install the versions in the correct location.

If there is already a way to do something like this that I am not aware of, I would like to hear about that as well! I searched the documentation and open issues but did not come across any solutions for this except for involving the --deployment flag but I want to avoid tracking dependencies in Git.

Thanks!

---

Solutions for problems I can forsee:

Q: How can I automatically bump minor/patch versions when everything in my lock file is a single version?
A: `bundle outdated` should be able to list versions that have minor or patch updates. Running a `bundle update <packages>` should read from the Gemfile to determine the version ranges and update them accordingly.

**Comment by segiddins** [comment]
`bundle install` already respects what's in the lockfile if you haven't changed your gemfile, and you can enforce that with `bundle install --deployment`, see http://bundler.io/v1.15/deploying.html for information about deploying gems with Bundler

**Comment by JoelLarson** [comment]
`bundle install` does respect what is in the lock file, but it does not lock exact versions and just locks what is in the `Gemfile` of your project and the dependencies within it.

`bundle install --deployment` also does not suit a build server well because it means that you would have to either commit your vendor folder to your code repository or package the whole thing as an asset from a development machine.

**Comment by segiddins** [comment]
> but it does not lock exact versions and just locks what is in the Gemfile of your project and the dependencies within it

That's not true, it _does_ lock exact versions

See this lockfile, for example:

```
GIT
  remote: https://github.com/cocoapods/molinillo.git
  revision: c2f655cd697812574a8537cc7c9cf68a0c0357b3
  specs:
    molinillo (0.5.7)

GEM
  remote: https://rubygems.org/
  specs:
    actioncable (5.1.2)
      actionpack (= 5.1.2)
      nio4r (~> 2.0)
      websocket-driver (~> 0.6.1)
    actionmailer (5.1.2)
      actionpack (= 5.1.2)
      actionview (= 5.1.2)
      activejob (= 5.1.2)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 2.0)
    actionpack (5.1.2)
      actionview (= 5.1.2)
      activesupport (= 5.1.2)
      rack (~> 2.0)
      rack-test (~> 0.6.3)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.1.2)
      activesupport (= 5.1.2)
      builder (~> 3.1)
      erubi (~> 1.4)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.3)
    activejob (5.1.2)
      activesupport (= 5.1.2)
      globalid (>= 0.3.6)
    activemodel (5.1.2)
      activesupport (= 5.1.2)
    activerecord (5.1.2)
      activemodel (= 5.1.2)
      activesupport (= 5.1.2)
      arel (~> 8.0)
    activesupport (5.1.2)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (~> 0.7)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    arel (8.0.0)
    builder (3.2.3)
    concurrent-ruby (1.0.5)
    erubi (1.6.1)
    globalid (0.4.0)
      activesupport (>= 4.2.0)
    i18n (0.8.6)
    loofah (2.0.3)
      nokogiri (>= 1.5.9)
    mail (2.6.6)
      mime-types (>= 1.16, < 4)
    method_source (0.8.2)
    mime-types (3.1)
      mime-types-data (~> 3.2015)
    mime-types-data (3.2016.0521)
    mini_portile2 (2.2.0)
    minitest (5.10.2)
    nio4r (2.1.0)
    nokogiri (1.8.0)
      mini_portile2 (~> 2.2.0)
    rack (2.0.3)
    rack-test (0.6.3)
      rack (>= 1.0)
    rails (5.1.2)
      actioncable (= 5.1.2)
      actionmailer (= 5.1.2)
      actionpack (= 5.1.2)
      actionview (= 5.1.2)
      activejob (= 5.1.2)
      activemodel (= 5.1.2)
      activerecord (= 5.1.2)
      activesupport (= 5.1.2)
      bundler (>= 1.3.0, < 2.0)
      railties (= 5.1.2)
      sprockets-rails (>= 2.0.0)
    rails-dom-testing (2.0.3)
      activesupport (>= 4.2.0)
      nokogiri (>= 1.6)
    rails-html-sanitizer (1.0.3)
      loofah (~> 2.0)
    railties (5.1.2)
      actionpack (= 5.1.2)
      activesupport (= 5.1.2)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (12.0.0)
    sprockets (3.7.1)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.2.0)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
    thor (0.19.4)
    thread_safe (0.3.6)
    tzinfo (1.2.3)
      thread_safe (~> 0.1)
    websocket-driver (0.6.5)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.2)

PLATFORMS
  ruby

DEPENDENCIES
  molinillo!
  rails

BUNDLED WITH
   1.15.1
```

The lines `gem_name (gem_version)` lock the exact version

**Comment by indirect** [comment]
Thanks for writing up a detailed description of your issue!

I think it might help to clarify a few things:

1. Bundler does not use Gemfile or Gemfile.lock files from dependencies. Ever.
2. Bundler puts the exact version number or git sha in the lock and only ever installs that exact version, ever.
3. Bundler's deployment mode does not create or interact with the vendor directory in any way. Instead, it raises an exception if the Gemfile has changed and no longer matches the lock file.

In conclusion, Bundler already does exactly what you have asked for, so I am marking this ticket as resolved.

**Comment by JoelLarson** [comment]
Thanks for the explanation segiddins and indirect! I see that my original assessment was incorrect.
